src/providers/twitch/IrcMessageHandler.cpp

```diff
@@ -466,43 +467,73 @@ void IrcMessageHandler::populateReply(
+            // Linear search is justified by the infrequent use of replies
+            for (auto &otherMsg : otherLoaded)
             {
-                // Found root reply message
-                foundMessage = otherMsg;
-                break;
+                if (otherMsg->id == replyID)
+                {
+                    // Found root reply message
+                    foundMessage = otherMsg;
+                    break;
+                }
             }
-        }
 
-        if (!foundMessage)
-        {
-            // We didn't find the reply root message in the otherLoaded messages
-            // which are typically the already-parsed recent messages from the
-            // Recent Messages API. We could have a really old message that
-            // still exists being replied to, so check for that here.
-            foundMessage = channel->findMessage(replyID);
+            if (!foundMessage)
+            {
+                // We didn't find the reply root message in the otherLoaded messages
+                // which are typically the already-parsed recent messages from the
+                // Recent Messages API. We could have a really old message that
+                // still exists being replied to, so check for that here.
+                foundMessage = channel->findMessage(replyID);
+            }
+
+            if (foundMessage)
+            {
+                std::shared_ptr<MessageThread> newThread =
+                        std::make_shared<MessageThread>(foundMessage);
+                updateReplyParticipatedStatus(tags, message->nick(), builder,
+                                              newThread, true);
+
+                builder.setThread(newThread);
+                rootThread = newThread;
+                // Store weak reference to thread in channel
+                channel->addReplyThread(newThread);
+            }
         }
 
-        if (foundMessage)
+        if (const auto parentIt = tags.find("reply-parent-msg-id"); parentIt != tags.end())
         {
-            std::shared_ptr<MessageThread> newThread =
-                std::make_shared<MessageThread>(foundMessage);
-            updateReplyParticipatedStatus(tags, message->nick(), builder,
-                                          newThread, true);
-
-            builder.setThread(newThread);
-            // Store weak reference to thread in channel
-            channel->addReplyThread(newThread);
+            const QString parentID = parentIt.value().toString();
+            std::shared_ptr<const Message> parent;
+            if (replyID == parentID)
+            {
+                if (rootThread)
+                {
+                    parent = rootThread->root();
+                }
+            }
+            else
+            {
+                auto parentThreadIt = channel->threads_.find(parentID);
+                if (parentThreadIt != channel->threads_.end() && !parentThreadIt->second.expired())
```

Comment:
```suggestion
                auto parentThreadIt = channel->threads().find(parentID);
                if (parentThreadIt != channel->threads().end() && !parentThreadIt->second.expired())
```

They're friends, but they don't need to be this close (the map could be in another variable, but I don't think that's necessary). Same for the change in `IrcMessageHandler::addMessage`.